### PR TITLE
Check Java classes are not used as objects

### LINKED ISSUE #972: Ycheck that java classes are not used as objects
This passes Ycheck, bu it should not:

```
import java.lang.Math

val myMath: Math.type = Math
myMath.sqrt(3.0)
```

### PR DESCRIPTION
**Pull request by liufengyun**
Following code should not pass Ycheck:

    import java.lang.Math

    val myMath: Math.type = Math
    myMath.sqrt(3.0)
    id(myMath)

Fix #972. Review @DarkDimius .

### CODE REVIEW

File: src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 1]
```diff
@@ -181,6 +181,106 @@ class TreeChecker extends Phase with SymTransformer {
...
+      case Match(selector, cases) =>
+        assertIdentNotJavaClass(selector)
+        cases.foreach(caseDef => assertIdentNotJavaClass(caseDef.body))
+      case Return(expr, _) =>
+        assertIdentNotJavaClass(expr)
+      case Try(expr, cases, finalizer) =>
+        assertIdentNotJavaClass(expr)
+        cases.foreach(caseDef => assertIdentNotJavaClass(caseDef.body))
+        assertIdentNotJavaClass(finalizer)
+      case untpd.Throw(expr) =>
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: No need to handle untyped trees here.
They would never reach Ycheck.

File: src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 2]
```diff
@@ -181,6 +181,106 @@ class TreeChecker extends Phase with SymTransformer {
...
 
+    /** assert Java classes are not used as objects */
+    def assertIdentNotJavaClass(tree: untpd.Tree)(implicit ctx: Context): Unit = tree match {
+      case _ : untpd.Ident =>
+        assert(!tree.symbol.is(JavaModule), "Java class can't be used as value: " + tree)
+      case _ =>
+    }
+
+    /** check Java classes are not used as objects */
+    def checkIdentNotJavaClass(tree: untpd.Tree)(implicit ctx: Context): Unit = tree match {
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: Ycheck checks typed trees. No need to support all the `untpd` stuff.

File: src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 3]
```diff
@@ -181,6 +181,106 @@ class TreeChecker extends Phase with SymTransformer {
...
+      // case tree: untpd.Ident =>
+      // case tree: untpd.Select =>
+      // case tree: untpd.SelectFromTypeTree =>
+      // case tree: untpd.Bind =>
+      case vd : ValDef =>
```
Comment (liufengyun) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: I keep the unchecked cases as comments to make it explicit what cases are ignored, in order to facilitate code maintenance. I'm not sure if it's a good style.

File: src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 4]
```diff
@@ -181,6 +181,106 @@ class TreeChecker extends Phase with SymTransformer {
...
+        assertIdentNotJavaClass(body)
+      // case untpd.GenFrom(_, expr) =>
+      case untpd.GenAlias(_, expr) =>
+        assertIdentNotJavaClass(expr)
+      // case untpd.ContextBounds(_, _) =>
+      case _ =>
+    }
+
+    override def typed(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): tpd.Tree = {
+      checkIdentNotJavaClass(tree)
```
Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: I would go with

```
checkIdentNotJavaClass(super.typed(tree))
```

As this pattern is commonly used in typer.

Comment (liufengyun) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: Thanks, I'll follow the approach.

BTW, it's safe to assume all untyped ASTs are desugared after `FrontEnd`?

Comment (DarkDimius) [inline src/dotty/tools/dotc/transform/TreeChecker.scala]: > BTW, it's safe to assume all untyped ASTs are desugared after FrontEnd?

yes.

**Comment by DarkDimius** [comment]
LGTM

**Comment by liufengyun** [comment]
Thanks @DarkDimius .

When you have time, could you please add some description to the benchmark infrastructure task #1210 ? That would help a lot for me to get started. Thanks a lot.

**Pull request merged**